Got functionality of allowing user to choose how many comments to see

casanted opened the pull request.

avaleske commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -24,11 +30,23 @@
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
```
Thanks for removing outdated comments!

avaleske commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -24,11 +30,23 @@
 public class DataServlet extends HttpServlet {
 
-  private List<String> comments;
+  class Comment {
```
This nested class is a really good idea.

avaleske commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -37,8 +55,25 @@ public void init() {
+    PreparedQuery results = datastore.prepare(query);
 
-    String json = convertToJson(comments);
+    for (Entity entity : results.asIterable()) {
```
I love for..of loops like this that don't need a loop index. I find them much easier to follow. :)

avaleske commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -37,8 +55,25 @@ public void init() {
-    String json = convertToJson(comments);
+    for (Entity entity : results.asIterable()) {
+      String name = (String) entity.getProperty("name");
+      String text = (String) entity.getProperty("comment");
```
What happens if the property isn't set? Or is it guaranteed to alway shave a property called "name"?

casanted replied:
In doPost, every property is set before it is stored in Datastore, and so I am thinking the code should not run into this issue. What do you think?

avaleske replied:
What if, in the future, you add a new property that old data doesn't have?

avaleske commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -49,11 +84,22 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
   }
 
-  private String convertToJson(List<String> comments) {
+  private String convertToJson(List<Comment> comments) {
```
Pulling this into its own function was a good way to hide boilerplate code and keep the main method clean. nice!

avaleske commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -37,8 +55,25 @@ public void init() {
+      comments.add(comment);
+    }
+
+    int limit = getLimit(request);
```
Two questions:
- It took me a second to understand what the limit was. Could we name this more clearly?
- It looks like we're getting all the comments out of the database, and then dropping anything other than the last `n`. That'll be slow if there's a lot of comments. Instead, could we update the query above based on the limit?

casanted replied:
I will rename the variable to a more suitable name. And yes I see that that is a more efficient way of doing it. I will try to implement that now.

avaleske commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -70,4 +116,25 @@ private String getParameter(HttpServletRequest request, String name, String defa
+      limit = Integer.parseInt(limitString);
+    } catch (NumberFormatException e) {
+      System.err.println("Could not convert to int: " + limitString);
+      return -1;
```
It's good that you're handling these errors, especially since someone could send a URL with any value in it!

Will returning -1 here cause problems for the `subList()` call that uses the result of this function?

casanted replied:
I think an if statement here will solve the issue, I will put that in the code as well

terencedignon commented:
Austin had some great comments, I'll take a look after he LGTMs. Going to place you as the assignee Carolyn, feel free to change this to Austin or I as you make revisions. Thanks!

casanted commented:
- Implemented the doGet and doPost functions for the /data servlet in order to store and print out comments properly  
- Changed the script.js to fetch the comments, depending on the limiting number the user chooses

avaleske commented:
Approving now because the fixes are in #8.
